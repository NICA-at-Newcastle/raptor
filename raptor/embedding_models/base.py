from typing import Generic, TypeVar, Protocol
import numpy as np

_CHUNK_contra = TypeVar("_CHUNK_contra", contravariant=True)


class IEmbeddingModel(Protocol, Generic[_CHUNK_contra]):
    """Defines the methods required for a raptor embedding model."""

    def create_embedding(self, chunk: _CHUNK_contra) -> np.ndarray: ...

    def create_text_embedding(self, text: str) -> np.ndarray: ...

    @property
    def slug(self) -> str: ...
