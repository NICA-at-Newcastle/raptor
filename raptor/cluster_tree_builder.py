from typing import override, Dict, List, TypeVar
import logging
import dataclasses
from concurrent.futures import ThreadPoolExecutor
from threading import Lock

from .tree_builder import TreeBuilder
from .utils import (
    get_node_list,
)
from .clustering_algorithms.base import ClusteringAlgorithm
from .clustering_algorithms.umap_gmm import UMAPGMMClusteringAlgorithm
from .tree_structures import Node

logging.basicConfig(format="%(asctime)s - %(message)s", level=logging.INFO)


_CHUNK = TypeVar("_CHUNK")
_C = TypeVar("_C")


class ClusterTreeBuilder(TreeBuilder[_CHUNK]):
    """Tree builder which uses clustering."""

    @dataclasses.dataclass
    class Config(TreeBuilder.Config[_C]):
        """ClusterTreeBuilder config"""

        clustering_algorithm: ClusteringAlgorithm = dataclasses.field(
            default_factory=lambda: UMAPGMMClusteringAlgorithm(
                UMAPGMMClusteringAlgorithm.Config()
            )
        )
        target_reduction_factor: float = dataclasses.field(default=2)

        def log_config(self):
            base_summary = super().log_config()
            cluster_tree_summary = f"""
            Clustering Algorithm: {self.clustering_algorithm}
            """
            return base_summary + cluster_tree_summary

    def __init__(self, config: Config[_CHUNK]) -> None:
        super().__init__(config)

        self.clustering_algorithm = config.clustering_algorithm
        self.target_reduction_factor = config.target_reduction_factor

        logging.info(
            "Successfully initialized ClusterTreeBuilder with Config %s",
            config.log_config(),
        )

    @override
    def construct_tree(
        self,
        current_level_nodes: Dict[int, Node],
        all_tree_nodes: Dict[int, Node],
        layer_to_nodes: Dict[int, List[Node]],
        use_multithreading: bool = False,
    ) -> tuple[Dict[int, Node], list[Node]]:
        logging.info("Using Cluster TreeBuilder")

        next_node_index = len(all_tree_nodes)
        all_outliers = []

        def process_cluster(
            cluster: list[Node],
            layer: int,
            new_level_nodes: dict[int, Node],
            next_node_index: int,
            lock: Lock,
        ):
            summarized_text = self.summarize(
                context=[node["chunk"] for node in cluster],
            )

            _, new_parent_node = self.create_node(
                next_node_index,
                summarized_text,
                layer,
            )

            for node in cluster:
                self.config.storage.save_node(
                    node,
                    next_node_index,
                )

            with lock:
                new_level_nodes[next_node_index] = new_parent_node

        for layer in range(self.config.num_layers):
            true_layer_number = layer + 1
            new_level_nodes = {}

            logging.info("Constructing Layer %s", true_layer_number)
            node_list_current_layer = get_node_list(current_level_nodes)

            logging.info("Clustering %s nodes", len(node_list_current_layer))
            clusters, outliers = self.clustering_algorithm(node_list_current_layer)
            logging.info("Got %s clusters", len(clusters))
            all_outliers.extend(outliers)

            reduction_factor = (
                0
                if len(clusters) == 0
                else len(node_list_current_layer) / len(clusters)
            )

            if reduction_factor < self.target_reduction_factor:
                self.num_layers = true_layer_number
                logging.info(
                    "Stopping Layer construction: Reduction factor of %s reached. Total Layers in tree: %s",
                    reduction_factor,
                    true_layer_number,
                )
                break

            lock = Lock()

            if use_multithreading:
                with ThreadPoolExecutor() as executor:
                    for cluster in clusters:
                        executor.submit(
                            process_cluster,
                            cluster,
                            true_layer_number,
                            new_level_nodes,
                            next_node_index,
                            lock,
                        )
                        next_node_index += 1
                    executor.shutdown(wait=True)

            else:
                for cluster in clusters:
                    process_cluster(
                        cluster,
                        true_layer_number,
                        new_level_nodes,
                        next_node_index,
                        lock,
                    )
                    next_node_index += 1

            layer_to_nodes[true_layer_number] = list(new_level_nodes.values())
            current_level_nodes = new_level_nodes
            all_tree_nodes.update(new_level_nodes)

        return current_level_nodes, all_outliers
