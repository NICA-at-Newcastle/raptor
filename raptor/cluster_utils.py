import logging
import random
from abc import abstractmethod
import dataclasses
from typing import List, Optional, override, TypeVar, Generic

import numpy as np
import umap
from sklearn.mixture import GaussianMixture
from .token_counter import BaseTokenCounter, BytePairTokenCounter
from .tree_structures import Node

# Initialize logging
logging.basicConfig(format="%(asctime)s - %(message)s", level=logging.INFO)


# Import necessary methods from other modules

# Set a random seed for reproducibility
RANDOM_SEED = 224
random.seed(RANDOM_SEED)


def global_cluster_embeddings(
    embeddings: np.ndarray,
    dim: int,
    n_neighbors: Optional[int] = None,
    metric: str = "cosine",
) -> np.ndarray:
    if n_neighbors is None:
        n_neighbors = int((len(embeddings) - 1) ** 0.5)
    reduced_embeddings = umap.UMAP(
        n_neighbors=n_neighbors, n_components=dim, metric=metric
    ).fit_transform(embeddings)
    assert isinstance(reduced_embeddings, np.ndarray)
    return reduced_embeddings


def local_cluster_embeddings(
    embeddings: np.ndarray, dim: int, num_neighbors: int = 10, metric: str = "cosine"
) -> np.ndarray:
    reduced_embeddings = umap.UMAP(
        n_neighbors=num_neighbors, n_components=dim, metric=metric
    ).fit_transform(embeddings)
    assert isinstance(reduced_embeddings, np.ndarray)
    return reduced_embeddings


def get_optimal_clusters(
    embeddings: np.ndarray, max_clusters: int = 50, random_state: int = RANDOM_SEED
) -> int:
    max_clusters = min(max_clusters, len(embeddings))
    n_clusters = np.arange(1, max_clusters)
    bics = []
    for n in n_clusters:
        gm = GaussianMixture(n_components=n, random_state=random_state)
        gm.fit(embeddings)
        bics.append(gm.bic(embeddings))
    optimal_clusters = n_clusters[np.argmin(bics)]
    return optimal_clusters


def GMM_cluster(embeddings: np.ndarray, threshold: float, random_state: int = 0):
    n_clusters = get_optimal_clusters(embeddings)
    gm = GaussianMixture(n_components=n_clusters, random_state=random_state)
    gm.fit(embeddings)
    probs = gm.predict_proba(embeddings)
    labels = [np.where(prob > threshold)[0] for prob in probs]
    return labels, n_clusters


def perform_clustering(
    embeddings: np.ndarray, dim: int, threshold: float, verbose: bool = False
) -> List[np.ndarray]:
    reduced_embeddings_global = global_cluster_embeddings(
        embeddings, min(dim, len(embeddings) - 2)
    )
    global_clusters, n_global_clusters = GMM_cluster(
        reduced_embeddings_global, threshold
    )

    if verbose:
        logging.info(f"Global Clusters: {n_global_clusters}")

    all_local_clusters = [np.array([]) for _ in range(len(embeddings))]
    total_clusters = 0

    for i in range(n_global_clusters):
        global_cluster_embeddings_ = embeddings[
            np.array([i in gc for gc in global_clusters])
        ]
        if verbose:
            logging.info(
                f"Nodes in Global Cluster {i}: {len(global_cluster_embeddings_)}"
            )
        if len(global_cluster_embeddings_) == 0:
            continue
        if len(global_cluster_embeddings_) <= dim + 1:
            local_clusters = [np.array([0]) for _ in global_cluster_embeddings_]
            n_local_clusters = 1
        else:
            reduced_embeddings_local = local_cluster_embeddings(
                global_cluster_embeddings_, dim
            )
            local_clusters, n_local_clusters = GMM_cluster(
                reduced_embeddings_local, threshold
            )

        if verbose:
            logging.info(f"Local Clusters in Global Cluster {i}: {n_local_clusters}")

        for j in range(n_local_clusters):
            local_cluster_embeddings_ = global_cluster_embeddings_[
                np.array([j in lc for lc in local_clusters])
            ]
            indices = np.where(
                (embeddings == local_cluster_embeddings_[:, None]).all(-1)
            )[1]
            for idx in indices:
                all_local_clusters[idx] = np.append(
                    all_local_clusters[idx], j + total_clusters
                )

        total_clusters += n_local_clusters

    if verbose:
        logging.info("Total Clusters: %s", total_clusters)
    return all_local_clusters


_CHUNK = TypeVar("_CHUNK")
_C = TypeVar("_C")


class ClusteringAlgorithm(Generic[_CHUNK]):
    """
    Base clustering algorithm. Clusters the given nodes into a list of clusters.
    """

    @abstractmethod
    def __call__(
        self,
        nodes: List[Node[_CHUNK]],
    ) -> list[list[Node[_CHUNK]]]:
        raise NotImplementedError("Implement in subclass")

    @property
    @abstractmethod
    def reduction_dimension(self) -> int:
        """Returns the reduction dimension of the clustering algorithm."""
        raise NotImplementedError("Implement in subclass")


class RAPTOR_Clustering(ClusteringAlgorithm[_CHUNK]):
    """Base Raptor clustering algorithm"""

    @dataclasses.dataclass
    class Config(Generic[_C]):
        """Raptor clustering algorithm config."""

        max_length_in_cluster: int = dataclasses.field(default=3500)
        token_counter: BaseTokenCounter = dataclasses.field(
            default_factory=BytePairTokenCounter
        )
        reduction_dimension: int = dataclasses.field(default=10)
        threshold: float = dataclasses.field(default=0.1)
        verbose: bool = dataclasses.field(default=False)

    def __init__(self, config: Config[_CHUNK]):
        self.config = config

    @override
    def __call__(
        self,
        nodes: List[Node[_CHUNK]],
    ) -> list[list[Node[_CHUNK]]]:
        # Get the embeddings from the nodes
        embeddings = np.array([node["embedding"] for node in nodes])

        # Perform the clustering
        clusters = perform_clustering(
            embeddings,
            dim=self.config.reduction_dimension,
            threshold=self.config.threshold,
        )

        # Initialize an empty list to store the clusters of nodes
        node_clusters = []

        # Iterate over each unique label in the clusters
        for label in np.unique(np.concatenate(clusters)):
            # Get the indices of the nodes that belong to this cluster
            indices = [i for i, cluster in enumerate(clusters) if label in cluster]

            # Add the corresponding nodes to the node_clusters list
            cluster_nodes = [nodes[i] for i in indices]

            # Base case: if the cluster only has one node, do not attempt to recluster it
            if len(cluster_nodes) == 1:
                node_clusters.append(cluster_nodes)
                continue

            # Calculate the total length of the text in the nodes
            total_length = sum(
                [self.config.token_counter(node["chunk"]) for node in cluster_nodes]
            )

            # If the total length exceeds the maximum allowed length, recluster this cluster
            if total_length > self.config.max_length_in_cluster:
                if self.config.verbose:
                    logging.info(
                        "reclustering cluster with %s nodes", len(cluster_nodes)
                    )
                node_clusters.extend(self(cluster_nodes))
            else:
                node_clusters.append(cluster_nodes)

        return node_clusters

    @property
    @override
    def reduction_dimension(self) -> int:
        return self.config.reduction_dimension
