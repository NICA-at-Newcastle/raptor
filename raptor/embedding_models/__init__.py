from .base import BaseEmbeddingModel


__all__ = ("BaseEmbeddingModel",)
