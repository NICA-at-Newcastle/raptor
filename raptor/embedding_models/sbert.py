import numpy as np
from sentence_transformers import SentenceTransformer


class SBertEmbeddingModel:

    def __init__(self, model_name="sentence-transformers/multi-qa-mpnet-base-cos-v1"):
        self._model_name = model_name
        self.model = SentenceTransformer(model_name)

    def create_embedding(self, text):
        embeddings = self.model.encode(text)
        assert isinstance(embeddings, np.ndarray)
        return embeddings

    @property
    def slug(self) -> str:
        return f"SBERT/{self._model_name}"
