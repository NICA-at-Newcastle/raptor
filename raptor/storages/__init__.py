from .base import IStorageSave, IStorageSearch
from .memory_storage import MemoryStorage


__all__ = (
    "IStorageSave",
    "IStorageSearch",
    "MemoryStorage",
)
