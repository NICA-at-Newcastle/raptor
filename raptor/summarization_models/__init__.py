from ._base import ISummarizationModel
from .debug import DebugSummarizationModel
from .bart import BartSummarizationModel
from .hugging_face import HFSummarizationModel
from .open_ai import (
    OpenAiSummarizationModel,
    GPT3SummarizationModel,
    GPT3TurboSummarizationModel,
)


__all__ = (
    "ISummarizationModel",
    "DebugSummarizationModel",
    "BartSummarizationModel",
    "HFSummarizationModel",
    "OpenAiSummarizationModel",
    "GPT3SummarizationModel",
    "GPT3TurboSummarizationModel",
)
