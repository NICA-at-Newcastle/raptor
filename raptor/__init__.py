from .cluster_tree_builder import ClusterTreeBuilder, ClusteringAlgorithm
from .tree_builder import TreeBuilder
from .tree_structures import Node, Tree
from .tree_retriever import TreeRetriever


__all__ = (
    "ClusterTreeBuilder",
    "ClusteringAlgorithm",
    "TreeBuilder",
    "TreeRetriever",
    "Node",
    "Tree",
)
