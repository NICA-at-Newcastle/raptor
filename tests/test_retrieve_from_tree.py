"""Tree retriever tests"""

from raptor import TreeRetriever
from raptor.embedding_models import SBertEmbeddingModel


def test_retreives_text_with_outliers(memory_storage, snapshot):
    """Test that tree raptor retrieves from pre-computed cinderella tree."""

    sbert = SBertEmbeddingModel()

    tree_retriever = TreeRetriever(
        TreeRetriever.Config(
            storage=memory_storage,
            embedding_model=sbert,
            limit=TreeRetriever.Limit.TopK(5),
        )
    )

    results = tree_retriever.retrieve(
        "What is cinderella doing?",
        tree_retriever.SearchMethod.Tree((-1, True)),
    )

    returned_chunks = [result["chunk"] for result in results]
    assert snapshot == returned_chunks


def test_retreives_text_no_outliers(memory_storage, snapshot):
    """Test that tree raptor retrieves from pre-computed cinderella tree."""

    sbert = SBertEmbeddingModel()

    tree_retriever = TreeRetriever(
        TreeRetriever.Config(
            storage=memory_storage,
            embedding_model=sbert,
            limit=TreeRetriever.Limit.TopK(5),
        )
    )

    results = tree_retriever.retrieve(
        "What is cinderella doing?",
        tree_retriever.SearchMethod.Tree((-1, False)),
    )

    returned_chunks = [result["chunk"] for result in results]
    assert snapshot == returned_chunks
