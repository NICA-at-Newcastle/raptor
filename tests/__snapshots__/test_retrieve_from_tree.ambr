# serializer version: 1
# name: test_retreives_text_no_outliers
  list([
  ])
# ---
# name: test_retreives_text_with_outliers
  list([
  ])
# ---
